File: packages/node/src/core/auth-user.ts
import { type RawClaims } from "./claims.js";

/**
 * The normalized, provider-agnostic representation of an authenticated principal.
 *
 * Every provider — Microsoft today, Google/Okta/Auth0 tomorrow — must map its
 * token into this shape. Consumers of the framework (route handlers, policies)
 * only ever see `AuthUser`; they never branch on which provider produced it.
 */
export interface AuthUser {
  /**
   * Stable, unique identifier for the principal within the issuing directory.
   * Providers map this from their most stable subject claim (Microsoft: `oid`,
   * falling back to `sub`).
   */
  readonly id: string;

  /** Primary email / UPN, if the token carries one. */
  readonly email?: string;

  /** Human-readable display name, if available. */
  readonly name?: string;

  /**
   * Application roles granted to the principal. Always an array (possibly empty)
   * so authorization checks never have to null-guard.
   */
  readonly roles: readonly string[];

  /**
   * OAuth2 scopes / delegated permissions present on the token. Always an array.
   */
  readonly scopes: readonly string[];

  /**
   * Tenant / organization identifier the principal belongs to, when the provider
   * is multi-tenant aware (Microsoft: `tid`).
   */
  readonly tenantId?: string;

  /** Name of the provider that authenticated this user (e.g. `"microsoft"`). */
  readonly provider: string;

  /**
   * The full set of validated claims from the token. Exposed (read-only) so
   * advanced consumers can read provider-specific claims without the framework
   * having to model every one of them.
   */
  readonly claims: Readonly<RawClaims>;
}

File: packages/go/core/auth_user.go
package core

// AuthUser is the normalized, provider-agnostic representation of an
// authenticated principal. Every provider — Microsoft today, Google/Okta/Auth0
// tomorrow — maps its token into this shape. Consumers only ever see AuthUser;
// they never branch on which provider produced it.
type AuthUser struct {
	// ID is a stable, unique identifier for the principal within the issuing
	// directory (Microsoft: oid, falling back to sub).
	ID string `json:"id"`

	// Email is the primary email / UPN, if the token carries one.
	Email string `json:"email,omitempty"`

	// Name is a human-readable display name, if available.
	Name string `json:"name,omitempty"`

	// Roles are the application roles granted to the principal. Never nil so
	// authorization checks never have to nil-guard.
	Roles []string `json:"roles"`

	// Scopes are the OAuth2 scopes / delegated permissions on the token. Never nil.
	Scopes []string `json:"scopes"`

	// TenantID is the tenant / organization identifier (Microsoft: tid), when the
	// provider is multi-tenant aware.
	TenantID string `json:"tenantId,omitempty"`

	// Provider is the name of the provider that authenticated this user.
	Provider string `json:"provider"`

	// Claims is the full set of validated claims from the token.
	Claims Claims `json:"claims"`
}

// HasRole reports whether the user was granted role.
func (u *AuthUser) HasRole(role string) bool {
	for _, r := range u.Roles {
		if r == role {
			return true
		}
	}
	return false
}

// HasScope reports whether the user's token carries scope.
func (u *AuthUser) HasScope(scope string) bool {
	for _, s := range u.Scopes {
		if s == scope {
			return true
		}
	}
	return false
}
